doc/environment-variables.md
# Environment Variables

PixivFE's behavior is controlled by environment variables. Currently, you can only set variables directly in your environment.

An example configuration is provided in [`.env.example`](https://codeberg.org/VnPower/PixivFE/src/branch/v2/.env.example).

This document is based on [`doc/Environment Variables.go`](https://codeberg.org/VnPower/PixivFE/src/branch/v2/doc/Environment%20Variables.go).

!!! tip
    To quickly set up PixivFE, you need to define two required environment variables:

    - `PIXIVFE_TOKEN`: Your Pixiv account cookie, which is necessary for accessing Pixiv's Ajax API. Refer to the [guide on obtaining the PIXIVFE_TOKEN cookie](obtaining-pixivfe-token.md) for details on how to acquire your Pixiv token.
    - `PIXIVFE_PORT`: The port number on which PixivFE will run, for example, `8282`.

    For basic usage, configure your environment variables as follows:
    ```
    PIXIVFE_TOKEN=123456_AaBbccDDeeFFggHHIiJjkkllmMnnooPP
    PIXIVFE_PORT=8282
    ```

    If you are setting up a development environment, enable the development mode by also setting:
    ```
    PIXIVFE_DEV=true
    ```

## Required variables

### `PIXIVFE_PORT` or `PIXIVFE_UNIXSOCKET`

**Required**: Yes (one of the two)

- `PIXIVFE_PORT`: Port to listen on, e.g., `PIXIVFE_PORT=8282`.
- `PIXIVFE_UNIXSOCKET`: [UNIX socket](https://en.wikipedia.org/wiki/Unix_domain_socket) to listen on, e.g., `PIXIVFE_UNIXSOCKET=/srv/http/pages/pixivfe`.

### `PIXIVFE_TOKEN`

**Required**: Yes

Your Pixiv account cookie, used by PixivFE for authorization to fully access Pixiv's Ajax API.

See the [Obtaining the `PIXIVFE_TOKEN` cookie](obtaining-pixivfe-token.md) guide for detailed instructions.

## Optional variables

### `PIXIVFE_HOST`

**Required**: No (ignored if `PIXIVFE_UNIXSOCKET` is set)

!!!note
    If you're **not using a reverse proxy** or **running PixivFE inside Docker**, you should set `PIXIVFE_HOST=0.0.0.0`. This will allow PixivFE to accept connections from any IP address or hostname. If you don't set this, PixivFE will refuse direct connections from other machines or devices on your network.

This setting specifies the hostname or IP address that PixivFE should listen on and accept incoming connections from. For example, if you want PixivFE to only accept connections from the same machine (your local computer), you can set `PIXIVFE_HOST=localhost`.

### `PIXIVFE_REQUESTLIMIT`

**Required**: No

Set to a number to enable the built-in rate limiter, e.g., `PIXIVFE_REQUESTLIMIT=15`.

It's recommended to enable rate limiting in the reverse proxy in front of PixivFE rather than using this.

### `PIXIVFE_IMAGEPROXY`

**Required**: No, defaults to using the built-in proxy

!!! note
    The protocol **must** be included in the URL, e.g., `https://piximg.example.com`, where `https://` is the protocol used.

The URL of the image proxy server. Pixiv requires `Referer: https://www.pixiv.net/` in the HTTP request headers to fetch images directly. Set this variable if you wish to use an external image proxy or are unable to get images directly from Pixiv.

See [hosting an image proxy server](hosting-image-proxy-server.md) or the [list of public image proxies](public-image-proxies.md).

### `PIXIVFE_USERAGENT`

**Required**: No

**Default:** `Mozilla/5.0 (Windows NT 10.0; rv:123.0) Gecko/20100101 Firefox/123.0`

The value of the `User-Agent` header used for requests to Pixiv's API.

### `PIXIVFE_ACCEPTLANGUAGE`

**Required**: No

**Default:** `en-US,en;q=0.5`

The value of the `Accept-Language` header used for requests to Pixiv's API. Change this to modify the response language.

### `PIXIVFE_PROXY_CHECK_INTERVAL`

**Required**: No

**Default:** `480`

The interval in minutes between proxy checks. Defaults to 480 minutes (8 hours) if not set.
You can disable this by setting the value to 0. Then, proxies will only be checked once at server initialization.

### `PIXIVFE_DEV`

**Required**: No

Set to any value to enable development mode, in which the server will live-reload HTML templates and disable caching, e.g., `PIXIVFE_DEV=true`.


doc/hosting/environment-variables.md
# Environment Variables

PixivFE's behavior is controlled by environment variables. Currently, you can only set variables directly in your environment.

An example configuration is provided in [`.env.example`](https://codeberg.org/VnPower/PixivFE/src/branch/v2/.env.example).

This document is based on [`doc/Environment Variables.go`](https://codeberg.org/VnPower/PixivFE/src/branch/v2/doc/Environment%20Variables.go).

!!! tip
    To quickly set up PixivFE, you need to define two required environment variables:

    - `PIXIVFE_TOKEN`: Your Pixiv account cookie, which is necessary for accessing Pixiv's Ajax API. Refer to the [guide on obtaining the PIXIVFE_TOKEN cookie](obtaining-pixivfe-token.md) for details on how to acquire your Pixiv token.
    - `PIXIVFE_PORT`: The port number on which PixivFE will run, for example, `8282`.

    For basic usage, configure your environment variables as follows:
    ```
    PIXIVFE_TOKEN=123456_AaBbccDDeeFFggHHIiJjkkllmMnnooPP
    PIXIVFE_PORT=8282
    ```

    If you are setting up a development environment, enable the development mode by also setting:
    ```
    PIXIVFE_DEV=true
    ```

## Required variables

### `PIXIVFE_PORT` or `PIXIVFE_UNIXSOCKET`

**Required**: Yes (one of the two)

- `PIXIVFE_PORT`: Port to listen on, e.g., `PIXIVFE_PORT=8282`.
- `PIXIVFE_UNIXSOCKET`: [UNIX socket](https://en.wikipedia.org/wiki/Unix_domain_socket) to listen on, e.g., `PIXIVFE_UNIXSOCKET=/srv/http/pages/pixivfe`.

### `PIXIVFE_TOKEN`

**Required**: Yes

Your Pixiv account cookie, used by PixivFE for authorization to fully access Pixiv's Ajax API.

See the [Obtaining the `PIXIVFE_TOKEN` cookie](obtaining-pixivfe-token.md) guide for detailed instructions.

## Optional variables

### `PIXIVFE_HOST`

**Required**: No (ignored if `PIXIVFE_UNIXSOCKET` is set)

!!!note
    If you're **not using a reverse proxy** or **running PixivFE inside Docker**, you should set `PIXIVFE_HOST=0.0.0.0`. This will allow PixivFE to accept connections from any IP address or hostname. If you don't set this, PixivFE will refuse direct connections from other machines or devices on your network.

This setting specifies the hostname or IP address that PixivFE should listen on and accept incoming connections from. For example, if you want PixivFE to only accept connections from the same machine (your local computer), you can set `PIXIVFE_HOST=localhost`.

### `PIXIVFE_REQUESTLIMIT`

**Required**: No

Set to a number to enable the built-in rate limiter, e.g., `PIXIVFE_REQUESTLIMIT=15`.

It's recommended to enable rate limiting in the reverse proxy in front of PixivFE rather than using this.

### `PIXIVFE_IMAGEPROXY`

**Required**: No, defaults to using the built-in proxy

!!! note
    The protocol **must** be included in the URL, e.g., `https://piximg.example.com`, where `https://` is the protocol used.

The URL of the image proxy server. Pixiv requires `Referer: https://www.pixiv.net/` in the HTTP request headers to fetch images directly. Set this variable if you wish to use an external image proxy or are unable to get images directly from Pixiv.

See [hosting an image proxy server](image-proxy-server.md) or the [list of public image proxies](../public-image-proxies.md).

### `PIXIVFE_USERAGENT`

**Required**: No

**Default:** `Mozilla/5.0 (Windows NT 10.0; rv:123.0) Gecko/20100101 Firefox/123.0`

The value of the `User-Agent` header used for requests to Pixiv's API.

### `PIXIVFE_ACCEPTLANGUAGE`

**Required**: No

**Default:** `en-US,en;q=0.5`

The value of the `Accept-Language` header used for requests to Pixiv's API. Change this to modify the response language.

### `PIXIVFE_PROXY_CHECK_INTERVAL`

**Required**: No

**Default:** `8h`

The interval in minutes between proxy checks. Defaults to 8 hours if not set.
Please specify this value in Go's `time.Duration` notation, e.g. `2h3m5s`.
You can disable this by setting the value to 0. Then, proxies will only be checked once at server initialization.

### `PIXIVFE_DEV`

**Required**: No

Set to any value to enable development mode, in which the server will live-reload HTML templates and disable caching, e.g., `PIXIVFE_DEV=true`.


config/config.go
// Global (Server-Wide) Settings

package config

import (
	"context"
	"errors"
	"log"
	"math/rand"
	"net/url"
	"strings"
	"time"

	"github.com/sethvargo/go-envconfig"
)

var GlobalConfig ServerConfig

type ServerConfig struct {
	Version      string
	StartingTime string // used in /about page

	Host string `env:"PIXIVFE_HOST"`

	// One of the two is required
	Port       string `env:"PIXIVFE_PORT"`
	UnixSocket string `env:"PIXIVFE_UNIXSOCKET"`

	Token          []string `env:"PIXIVFE_TOKEN,required"` // may be multiple tokens. delimiter is ','    maybe add some testing?
	InDevelopment  bool     `env:"PIXIVFE_DEV"`
	UserAgent      string   `env:"PIXIVFE_USERAGENT,overwrite"`
	AcceptLanguage string   `env:"PIXIVFE_ACCEPTLANGUAGE,overwrite"`
	RequestLimit   int      `env:"PIXIVFE_REQUESTLIMIT"` // if 0, request limit is disabled

	ProxyServer_staging string  `env:"PIXIVFE_IMAGEPROXY,overwrite"`
	ProxyServer         url.URL // proxy server URL, may or may not contain authority part of the URL

	ProxyCheckInterval time.Duration `env:"PIXIVFE_PROXY_CHECK_INTERVAL,overwrite"`
}

func (s *ServerConfig) LoadConfig() error {
	s.Version = "v2.8"
	log.Printf("PixivFE %s\n", s.Version)

	s.StartingTime = time.Now().UTC().Format("2006-01-02 15:04")

	// set default values with env:"...,overwrite"
	s.UserAgent = "Mozilla/5.0 (Windows NT 10.0; rv:123.0) Gecko/20100101 Firefox/123.0"
	s.AcceptLanguage = "en-US,en;q=0.5"
	s.ProxyServer_staging = BuiltinProxyUrl
	s.ProxyCheckInterval = 8 * time.Hour

	// load config from from env vars
	if err := envconfig.Process(context.Background(), s); err != nil {
		return err
	}

	if s.Port == "" && s.UnixSocket == "" {
		log.Fatalln("Either PIXIVFE_PORT or PIXIVFE_UNIXSOCKET has to be set.")
		return errors.New("Either PIXIVFE_PORT or PIXIVFE_UNIXSOCKET has to be set.")
	}

	{ // validate proxy server
		proxyUrl, err := url.Parse(s.ProxyServer_staging)
		if err != nil {
			panic(err)
		}
		s.ProxyServer = *proxyUrl
		if (proxyUrl.Scheme == "") != (proxyUrl.Host == "") {
			log.Panicf("proxy server url is weird: %s\nPlease specify e.g. https://example.com", proxyUrl.String())
		}
		if strings.HasSuffix(proxyUrl.Path, "/") {
			log.Panicf("proxy server path (%s) has cannot end in /: %s\nPixivFE does not support this now, sorry", proxyUrl.Path, proxyUrl.String())
		}
	}
	log.Printf("Proxy server set to: %s\n", s.ProxyServer.String())
	log.Printf("Proxy check interval set to: %v\n", s.ProxyCheckInterval)

	return nil
}

func GetRandomDefaultToken() string {
	defaultToken := GlobalConfig.Token[rand.Intn(len(GlobalConfig.Token))]

	return defaultToken
}
